integration_test/util_test.go

```diff
@@ -0,0 +1,361 @@
+			Denom:  "axpla",
+			Amount: feeAmt.Ceil().RoundInt(),
+		}
+
+		txhash, err := proposerWallet.SendTx(ChainID, proposalMsg, fee, xplaProposalGasLimit, false)
+		if txhash != "" && err == nil {
+			fmt.Println("Tx sent:", txhash)
+		} else {
+			return err
+		}
+
+		err = txCheck(txhash)
+		if err == nil {
+			fmt.Println("Tx applied", txhash)
+		} else {
+			return err
+		}
+
+		queryClient := txtypes.NewServiceClient(conn)
+		resp, err := queryClient.GetTx(context.Background(), &txtypes.GetTxRequest{
+			Hash: txhash,
+		})
+
+		if err != nil {
+			return err
+		}
+
+	PROPOSAL_RAISED:
+		for _, val := range resp.TxResponse.Events {
+			for _, attr := range val.Attributes {
+				if string(attr.Key) == "proposal_id" {
+					proposalId, _ = strconv.ParseUint(string(attr.Value), 10, 64)
+					break PROPOSAL_RAISED
+				}
+			}
+		}
+
+		fmt.Println("Proposal is applied as ID", proposalId)
+	}
+
+	{
+		fmt.Println("Voting...")
+
+		wg := sync.WaitGroup{}
+
+		errChan := make(chan error)
+		successChan := make(chan bool)
+
+		for _, addr := range voters {
+			wg.Add(1)
+
+			go func(addr *WalletInfo) {
+				defer wg.Done()
+
+				voteMsg := govtype.NewMsgVote(addr.ByteAddress, proposalId, govtype.OptionYes)
+				feeAmt := sdktypes.NewDec(xplaGeneralGasLimit).Mul(sdktypes.MustNewDecFromStr(xplaGasPrice))
+				fee := sdktypes.Coin{
+					Denom:  "axpla",
+					Amount: feeAmt.Ceil().RoundInt(),
+				}
+
+				txhash, err := addr.SendTx(ChainID, voteMsg, fee, xplaGeneralGasLimit, false)
+				if txhash != "" && err == nil {
+					fmt.Println(addr.StringAddress, "voted to the proposal", proposalId, "as tx", txhash, "err:", err)
+				} else {
+					errChan <- err
+					return
+				}
+
+				err = txCheck(txhash)
+				if err == nil {
+					fmt.Println(addr.StringAddress, "vote tx applied", txhash, "err:", err)
+				} else {
+					errChan <- err
+					return
+				}
+			}(addr)
+		}
+
+		go func() {
+			wg.Wait()
+			successChan <- true
+		}()
+
+	VOTE:
+		for {
+			select {
+			case chanErr := <-errChan:
+				return chanErr
+			case <-successChan:
+				break VOTE
+			}
+		}
+	}
+
+	fmt.Println("Waiting 25sec for the proposal passing...")
+	time.Sleep(time.Second * 25)
+	fmt.Println("Proposal tallied!")
+
+	return nil
+}
+
+func loadPrivValidator(validatorName string) (*PVKey, error) {
+	valKeyBytes, err := os.ReadFile(filepath.Join(".", validatorName, "priv_validator_key.json"))
+	if err != nil {
+		return nil, err
+	}
+
+	pvKey := PVKey{}
+	err = tmjson.Unmarshal(valKeyBytes, &pvKey)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pvKey, nil
+}
+
+func getValidatorList(in []stakingtype.Validator) []string {
+	ret := []string{}
+	for _, unit := range in {
+		ret = append(ret, unit.OperatorAddress)
+	}
+
+	return ret
+}
+
+func getMaxPrevoterCounts() (uint32, error) {
+	client, err := tmhttp.New("tcp://127.0.0.1:36657", "/websocket")
+	if err != nil {
+		return 0, err
+	}
+
+	resp, err := client.ConsensusState(context.Background())
```

Comment:
This is good, but I think it would be more accurate to check the block information.
If we search the block, we can check the validator address that it voted for.

Reply:
https://github.com/xpladev/xpla/pull/84/commits/a60b0878bc1fb21c03e9942889b74d601f11dd4a